Fix: RL base feature parity

NanoCode012 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

<!--- Describe your changes in detail -->

RL trainer was not loading some basic configs like `logging_steps` etc. This PR consolidates the setting of these params and cleans them up.

This PR also fixes a case where we did not call `.map` with `num_proc`.

To discuss:
- [x] Handling of bf16/bfloat16
- [x] Handling of fp16 in RL
- [x] Handling of tf32 in RL
- [x] Default of `use_reentrant`
- [x] Default of `remove_unused_columns`

Addresses a lot of points in #2105 

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

~Untested!~

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

## Social Handles (Optional)

<!-- Thanks for submitting a bugfix or enhancement. -->
<!-- We'd love to show our thanks to you on Twitter & Discord if you provide your handle -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added new trainer builder modules for both causal language modeling and reinforcement learning workflows, offering more modular and extensible configuration options.
  - Introduced support for a new configuration parameter `torch_compile_mode` and an optional field `dpo_label_smoothing` for advanced training customization.

- **Refactor**
  - Restructured trainer builder logic into dedicated modules, improving maintainability and clarity.
  - Updated import paths throughout the codebase to reflect the new modular trainer builder structure.
  - Enhanced optimizer integration for trainers using mixins for more flexible initialization and configuration.

- **Bug Fixes**
  - Corrected reward function handling and argument blocklisting in GRPO strategy to improve reliability.

- **Documentation**
  - Updated documentation configuration to reflect the new builder module structure and clarified certain configuration options.

- **Tests**
  - Added comprehensive unit and integration tests for the new trainer builder modules, covering various training configurations and optimizer scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

NanoCode012 committed:
feat: add num_proc and load from cache for rl mapping

NanoCode012 committed:
fix: refactor sft and rl trainer to set same base args

NanoCode012 committed:
feat: add report_to to set run name

NanoCode012 committed:
fix: consolidate handling of fp16, bf16, tf32 kwarg

NanoCode012 committed:
chore: consolidate eval_strat, loraplus, lr sched, max_length

NanoCode012 committed:
fix: deprecate old types

NanoCode012 committed:
fix: adding missing Any

NanoCode012 committed:
fix: max_steps incorrectly set

NanoCode012 committed:
fix: remove unnecessary datacollator kwarg insert and pop

NanoCode012 committed:
fix: update default max_steps

bursteratom reviewed (approved):
LGTM

NanoCode012 committed:
fix: add missing weight_decay handling

NanoCode012 committed:
fix: ignore max_length for grpo

NanoCode012 committed:
feat: update CI on trainer_builder

NanoCode012 committed:
fix: comments

winglian commented:
Thanks @NanoCode012 ! This should be good to go once the multi gpu tests pass too https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13457763772

winglian commented:
latest ci: https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13469023440

winglian commented:
looks like the multi-gpu GRPO tests are failing https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13469023440/job/37640091555

winglian commented on src/axolotl/core/trainer_builder.py:
```diff
@@ -224,6 +224,155 @@ def hook_post_create_trainer(self, trainer):
+        logging_steps = (
+            self.cfg.logging_steps
+            if self.cfg.logging_steps is not None
+            else max(min(int(0.005 * total_num_steps), 10), 1)
```
for many of the RL trainers, we can't rely on total_num_steps (set to None) and we pass -1 as max_steps to let the trainer figure out the total number of steps

winglian commented on src/axolotl/core/trainer_builder.py:
```diff
@@ -224,6 +224,155 @@ def hook_post_create_trainer(self, trainer):
+        else:
+            warmup_steps = min(int(0.03 * total_num_steps), 100)
+        if warmup_steps == 1:
+            warmup_steps = 2
```
what is this for?

NanoCode012 replied:
This was in the original SFT trainer https://github.com/axolotl-ai-cloud/axolotl/pull/2133/files#diff-5edc13801ecfdd108e81872527bdc78c6d24a73833968147b0a9ecb8452996f4L323-L324

winglian committed:
improve handling of warmup/logging steps

winglian commented:
https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13485689047

winglian committed:
use transformers default for logging steps, not None

NanoCode012 committed:
fix: remove redundant override

NanoCode012 committed:
fix: lint

NanoCode012 committed:
feat: allow custom optim for rl methods

NanoCode012 committed:
fix: duplicate optim setting

NanoCode012 committed:
fix(test): set sequence_parallel_degree default in base cfg

djsaunde commented:
One suggestion: can we break out the `trainer_builder.py` file into smaller modules? E.g., something like:

```shell
builders
├── __init__.py
├── base.py
├── causal.py
└── rl.py
```

NanoCode012 commented:
> One suggestion: can we break out the `trainer_builder.py` file into smaller modules? E.g., something like:
> 
> ```shell
> builders
> ├── __init__.py
> ├── base.py
> ├── causal.py
> └── rl.py
> ```

Yep, I can do that tomorrow!

NanoCode012 committed:
feat: add handling for seed and SP/ring-attn config

NanoCode012 committed:
chore: add back return typing from rebase

NanoCode012 committed:
fix(test): use RLType directly to skip needing to validate

NanoCode012 committed:
feat: split training builder into sub modules

NanoCode012 committed:
fix: remove deprecated clause

NanoCode012 committed:
chore: add missing config to doc

NanoCode012 committed:
fix: update quarto autodoc

NanoCode012 committed:
fix: import path for trainer builder and submodules

NanoCode012 committed:
fix: remove redundant configs from rebase mistake

NanoCode012 committed:
chore: simplify dynamo check

NanoCode012 committed:
fix: optimizer_cls_and_kwargs to be passed into trainer_kwargs

NanoCode012 committed:
fix: add missing rex from rebase

NanoCode012 committed:
fix: move pop optimizer_cls_and_kwargs

NanoCode012 committed:
fix: pop optimizer cls in rl too

NanoCode012 committed:
fix: leftover bug from rebase

NanoCode012 committed:
fix: update handling of trainer_cls in RL

NanoCode012 committed:
fix: address pr feedback

NanoCode012 committed:
feat: call hook_pre_create_trainer for rl

NanoCode012 committed:
chore: lint

NanoCode012 committed:
fix: return notimplemented for ppo

NanoCode012 committed:
feat: moved torch compile to base and refactor collator setting

NanoCode012 committed:
chore: remove unused importlib.util import

NanoCode012 committed:
fix: optimizer cls not being popped

NanoCode012 committed:
feat: move epoch setting to base

NanoCode012 committed:
fix: catch unhandled custom optimizer

NanoCode012 committed:
fix: remove duplicate lora plus setting

NanoCode012 committed:
chore: refactor if condition

NanoCode012 committed:
chore: refactor set_base_training_args into smaller modules

NanoCode012 committed:
fix: address TrainerBuilderBase class variables to instance var

NanoCode012 committed:
fix: add handling for beta3 and episilon2

NanoCode012 committed:
fix: change to pass dict via arg instead of updating dict

NanoCode012 committed:
Merge branch 'main' into fix/orpo_feature_parity

NanoCode012 commented:
Since the PR is quite old, rebasing proved to be a bit challenging, so I'm swapping to merge from main.

I think this PR is ready to be re-reviewed / almost good to go.

TODO left:
- [ ] whether to set `eval_accumulation_steps` to grad accu in RL (like in SFT)

NanoCode012 committed:
chore: simplify if condition

NanoCode012 committed:
fix: force access to lr & weight decay in case not provided to early error

NanoCode012 committed:
fix: remove log sweep

NanoCode012 committed:
Merge branch 'main' into fix/orpo_feature_parity

NanoCode012 committed:
chore: refactor if condition

NanoCode012 committed:
fix: address renamed cfg

NanoCode012 committed:
fix: improve handling of cosine hyp

djsaunde commented on _quarto.yml:
```diff
@@ -17,7 +17,9 @@ quartodoc:
         - convert
         - prompt_tokenizers
         - logging_config
-        - core.trainer_builder
```
I kind of prefer `core.builders.base`, etc. naming, what do you think?

NanoCode012 replied:
I'm open to this. Would we need to adjust the class to be named BuilderBase from TrainerBuilderBase? It may be a bit ambiguous what Builder refers to.

djsaunde replied:
IMO you should keep the naming as is

djsaunde commented on docs/config.qmd:
```diff
@@ -529,7 +530,7 @@ profiler_steps: # enable the pytorch profiler to capture the first N steps of tr
 loss_watchdog_threshold: # High loss value, indicating the learning has broken down (a good estimate is ~2 times the loss at the start of training)
 loss_watchdog_patience: # Number of high-loss steps in a row before the trainer aborts (default: 3)
 
-# Save model as safetensors (require safetensors package)
```
What does "Transformers default True" mean?

NanoCode012 replied:
It means the default upstream is true. I'll update it to be clearer

djsaunde commented on src/axolotl/core/trainer_builder/__init__.py:
nit: can you import the classes you intend to expose here and put them in `__all__`? Then, you can shorten the import statement elsewhere:

```python
from axolotl.core.trainer_builder import ...
```

NanoCode012 replied:
Yup, will make this change!

djsaunde commented on src/axolotl/core/trainer_builder/base.py:
```diff
@@ -0,0 +1,496 @@
+        else:
+            training_args_kwargs["bf16"] = self.cfg.bf16 or self.cfg.bfloat16
+
+    def _configure_optimizer_and_scheduler(self, training_args_kwargs):
```
IMO you could break this out into two methods, this is pretty large.

NanoCode012 replied:
I split scheduler out, but not sure if that's enough

djsaunde commented on src/axolotl/core/trainer_builder/base.py:
```diff
@@ -0,0 +1,496 @@
+
+            # Adam-specific kwargs
+            adam_kwargs: dict = {}
+            if training_args_kwargs.get("adam_beta1") and training_args_kwargs.get(
```
why can't the user specify one of these params and not the other?

NanoCode012 replied:
Hm, `betas` require a list of beta1 and beta2. We can certainly check for beta1, but if beta2 exist but not beta1, I'm not sure if we should have a default for this.

djsaunde commented on src/axolotl/core/trainer_builder/base.py:
```diff
@@ -0,0 +1,496 @@
+
+        self._configure_save_and_eval_strategy(training_args_kwargs)
+
+        self._configure_gradient_checkpointing(training_args_kwargs)
```
personally I prefer:

```suggestion
        self._configure_warmup_and_logging(total_num_steps, training_args_kwargs)
        self._configure_precision_settings(training_args_kwargs)
        self._configure_save_and_eval_strategy(training_args_kwargs)
        self._configure_gradient_checkpointing(training_args_kwargs)
```

djsaunde commented on src/axolotl/core/trainer_builder/base.py:
```diff
@@ -0,0 +1,496 @@
+        self._configure_gradient_checkpointing(training_args_kwargs)
+
+        # set arg into trainer_args_kwargs with same name if value not None
+        for arg in [
```
IMO this should be a constant

NanoCode012 replied:
I don't think it would be reused anywhere else, so would a constant be useful?

djsaunde commented on src/axolotl/core/trainer_builder/base.py:
```diff
@@ -0,0 +1,496 @@
+
+        self._configure_optimizer_and_scheduler(training_args_kwargs)
+
+        self._configure_torch_compile(training_args_kwargs)
```
```suggestion
        self._configure_reporting(training_args_kwargs)
        self._configure_hub_parameters(training_args_kwargs)
        self._configure_optimizer_and_scheduler(training_args_kwargs)
        self._configure_torch_compile(training_args_kwargs)
```

djsaunde commented on src/axolotl/core/trainer_builder/base.py:
```diff
@@ -0,0 +1,496 @@
+
+        training_args_kwargs["max_steps"] = self.cfg.max_steps or total_num_steps or -1
+
+        training_args_kwargs["num_train_epochs"] = self.cfg.num_epochs
```
```suggestion
        training_args_kwargs["max_steps"] = self.cfg.max_steps or total_num_steps or -1
        training_args_kwargs["num_train_epochs"] = self.cfg.num_epochs
```

djsaunde commented on src/axolotl/core/trainer_builder/rl.py:
```diff
@@ -0,0 +1,246 @@
+
+        return trainer_cls, trainer_cls_args
+
+    def build_training_arguments(self, total_num_steps):
```
Could be private

djsaunde commented on src/axolotl/core/trainer_builder/rl.py:
```diff
@@ -0,0 +1,246 @@
+        if self.cfg.use_wandb and training_args.run_name == training_args.output_dir:
+            training_args.run_name = (  # pylint: disable=attribute-defined-outside-init
+                None
+            )
```
What does this do exactly?

NanoCode012 replied:
I think, if you run experiment with RLHF previously to wandb, it would set the run name to `outputs/lora` etc . This would unset the `run_name` so that wandb would random the name `{random_animal_name}`.

I think this may be because Trainer internally sets `run_name={output_dir}` if not provided.

djsaunde commented on src/axolotl/core/trainer_builder/rl.py:
```diff
@@ -0,0 +1,246 @@
+        return trainer
+
+
+class HFPPOTrainerBuilder(TrainerBuilderBase):
```
Don't think this is used anywhere

NanoCode012 replied:
We also don't use `TRLPPOTrainer` but I'll leave this for a separate PR to discuss remove these.

djsaunde commented on src/axolotl/core/trainer_builder/sft.py:
```diff
@@ -0,0 +1,490 @@
+    using TRL.
+    """
+
+    def get_callbacks(self):
```
can be private

NanoCode012 replied:
If the child class calls, `super()._get_callbacks()` , would this make linter complain accessing private attribute? I was mid-refactoring the other functions when I got the above thought.

djsaunde commented on src/axolotl/core/trainer_builder/sft.py:
```diff
@@ -0,0 +1,490 @@
+
+        return callbacks
+
+    def get_post_trainer_create_callbacks(self, trainer):
```
can be private

djsaunde commented on src/axolotl/core/trainer_builder/sft.py:
```diff
@@ -0,0 +1,490 @@
+
+        return trainer
+
+    def build_collator(
```
could be private

djsaunde commented on src/axolotl/core/trainers/grpo/__init__.py:
```diff
@@ -167,4 +167,4 @@ def get_reward_func(cls, reward_func_fqn: str) -> RewardFunc:
                 f"Reward function {reward_func_fqn} is a pre-trained model path - if this is unexpected, please check the reward function path."
             )
-            return reward_func
+            return reward_func_fqn
```
fqn -> fn?

NanoCode012 replied:
This is intended. If the attempt to load the `reward_func` failed, then we assume `reward_func_fqn` is the valid path

djsaunde reviewed (commented):
Great stuff!

Mostly low impact comments from me.

NanoCode012 committed:
fix: remove unused params

salmanmohammadi commented on src/axolotl/core/trainer_builder/rl.py:
```diff
@@ -0,0 +1,246 @@
+            if trainer_cls is not None:
+                return trainer_cls, trainer_cls_args
+
+        if self.cfg.rl is RLType.GRPO:
```
nit: set `trainer_cls_args = [self.model]` once and extend/append as appropriate

NanoCode012 committed:
chore: refactor

NanoCode012 committed:
chore: clarify doc safetensors

NanoCode012 committed:
fix: update import path to be unified following comments

NanoCode012 committed:
fix: duplicate kwargs passed

NanoCode012 committed:
feat: return separate trainer_kwargs

NanoCode012 committed:
chore: refactor

NanoCode012 committed:
chore: refactor based on comments

NanoCode012 committed:
chore: refactor based on comments

NanoCode012 committed:
fix: move gpustats callback to base

NanoCode012 committed:
chore: create trainer_cls_args first based on comments

NanoCode012 committed:
fix: ipo label smoothing passed incorrectly

NanoCode012 committed:
feat: add optimizer parity for RL methods with test

NanoCode012 committed:
feat: add parity for optimizer in RM/PRM and add test

NanoCode012 committed:
fix: remove redundant function override for orpo/cpo batch metrics

NanoCode012 committed:
Merge branch 'main' into fix/orpo_feature_parity

NanoCode012 committed:
fix: improve handling of dpo_label_smoothing and merge issue

NanoCode012 committed:
fix: test fixture returning wrong field

NanoCode012 commented:
In https://github.com/axolotl-ai-cloud/axolotl/pull/2133/commits/e07c33c5603afa1fade6d49b539ea93918c0e784 and https://github.com/axolotl-ai-cloud/axolotl/pull/2133/commits/0e52b4d68bfadd76484701b0f2dab351bc15a52b , I have added parity for setting custom optimizer in RL and RM/PRM.

I think I addressed most of the comments left as well with one under discussion for rename from `trainer_builder->builders` and refactoring to private functions within the builders.

NanoCode012 committed:
fix: address avoid direct modify fixture

NanoCode012 committed:
chore: minor refactor

NanoCode012 committed:
Revert "chore: refactor"

This reverts commit 99c8859eb02d1c4bde465d1fb0290a8024d2fd87.

djsaunde reviewed (approved):
IMO good to go once the aforementioned naming change is in. Great stuff!

NanoCode012 committed:
Merge branch 'main' into fix/orpo_feature_parity

NanoCode012 committed:
feat: rename trainer_builder to builders

NanoCode012 commented:
Thanks for all the reviews. It was big PR. If there are any missing gaps, I'll address them in a follow up PR.
